java/runtime/src/main/java/io/ray/runtime/object/ObjectRefImpl.java

```diff
@@ -53,4 +51,53 @@ public ObjectId getId() {
     return type;
   }
 
+  @Override
+  protected void finalize() throws Throwable {
+    try {
+      removeLocalReference();
+    } finally {
+      super.finalize();
+    }
+  }
+
+  static ThreadLocal<Set<ObjectId>> containedObjectIds = ThreadLocal.withInitial(HashSet::new);
+
+  public static List<ObjectId> getAndClearContainedObjectIds() {
+    List<ObjectId> ids = new ArrayList<>(containedObjectIds.get());
+    containedObjectIds.get().clear();
+    return ids;
+  }
+
+  @Override
+  public void writeExternal(ObjectOutput out) throws IOException {
+    out.writeObject(this.getId());
+    out.writeObject(this.getType());
+    containedObjectIds.get().add(this.getId());
```

Comment:
Sorry I think I'm just not familiar with Java, but what do these methods do? I'm a bit confused about the last line for adding to the containedObjectIds. It looks like the object is adding itself, but shouldn't it be adding to the object that it's contained in?

Reply:
The `containedObjectIds` is a thread local static field. It's used to record all the nested object IDs of a certain object, like `self._thread_local.object_refs` in `serialization.py`.

When an outer object is being serialized, the nested `ObjectRef`s are all serialized and the `writeExternal` method of the nested `ObjectRef`s are executed. So after the outer object is serialized, the `containedObjectIds` field will contain all the nested object IDs.

This is focusing maybe because the `containedObjectIds` is defined in the `ObjectRefImpl` class. I'll move it to the `ObjectSerializer` class. Hope this can help.